rfcs/029-work_state_modelling/README.md

```diff
@@ -0,0 +1,177 @@
+
+There are some upcoming use cases where we need to modify the work modelling, such as adding relations (`parts`, `partOf`, `precededBy`, `succeededBy`) to the model for works after denormalisation by the relation embedder. Also @jamesgorrie and @jtweed have been investigating having additional sorts of works in the type hierarchy (such as collections, series, items) similarly to how there are different sorts of concepts.
+
+## Problem
+
+There are a number of issues with the way the modelling is currently implemented, primarily to do with composability of the types and the fact that they do not express our business logic as clearly as they could:
+
+1) To add a new sort of work we need to add a large amount of subtypes. For example to add a work such as a `Collection` we would potentially need the addition of up to 6 new types: `UnidentifiedCollection`, `UnidentifiedInvisibleCollection` etc.
+
+2) There is no way to add new state dependent data in a type safe manner without also adding more types of work for each case. To take the relation embedder as an example, that would require new types such as `DenormalisedWork`, `DenormalisedCollection` etc.
+
+3) There are already 2 type parameters on `WorkData`, both for the identified state of things such as concepts, and also for the identified state of images. Also, if we decided to add relation data directly to `WorkData` (as optional fields) rather than creating new work types we would need an additional third type parameter (to express whether the referenced works are pre or post minter). These are all tied together so ideally we should be able to specify them with a single type parameter.
+
+4) The type hierarchy from `BaseWork` downwards is quite complex and it can be hard to understand the meaning of particular types of work within the code. The naming could also do with changes in places (e.g. `TransformedBaseWork` does not make much sense as a name, given that the concept of a work does not exist before the transformer).
+
+## Proposed Solution
+
+This RFC proposes having a cleaner separation of the pipeline state of a work with the sort of work it is, namely with different forms of polymorphism:
+
+1) Sub-typing is used for differentiating different sorts of work, with a `Work` being a sum type ADT consisting of the finite number of possible case classes. Note this proposed sub-typing is much simpler than the current implementation, having only a single level of depth and a single parent type.
+
+2) Parameterisation of the `Work` by a state parameter indicates what stage of the pipeline it is in, and which can hold specific data depending on stage. By constraining the parameter to one of a set of known types, we can consider the `Work` model as a finite state machine with the stages in our pipeline each containing a particular transition.
```

Comment:
Nice application of theory to reality. It will be possible to implement this in such a way that the allowed transitions (and hence the progression of the Work type) are represented in the code, which will be _great_